infrastructure/src/main/java/org/corfudb/infrastructure/redundancy/RedundancyCalculator.java

```diff
@@ -0,0 +1,218 @@
+     * closer to the beginning or to the end of a layout segment list.
+     *
+     * @param layout Current layout.
+     * @param server Current server.
+     * @return True, if the server is not present in any segment.
+     */
+    public static boolean canRestoreRedundancy(Layout layout, String server) {
+        if (layout.getSegments().size() == 1) {
+            return false;
+        } else {
+
+            IntPredicate nodeNotPresent = currentIndex -> Sets.difference(
+                    layout.getSegments().get(currentIndex).getAllLogServers(),
+                    layout.getSegments().get(currentIndex - 1).getAllLogServers()
+            ).contains(server);
+
+            return IntStream.range(1, layout.getSegments().size())
+                    .boxed()
+                    .reduce(false,
+                            (nodePresentSoFar, currentIndex) ->
+                                    nodePresentSoFar || nodeNotPresent.test(currentIndex),
+                            (nodePresentBefore, nodePresentNow) ->
+                                    nodePresentBefore || nodePresentNow);
+        }
+    }
+
+    /**
+     * Returns true if any server can merge the segments of this layout.
+     * Any server can merge the segments of this layout if the segment is split
+     * and a set difference of log servers between the first two adjacent segments is empty.
+     *
+     * @param layout Current layout.
+     * @return True if the segments can be merged.
+     */
+    public static boolean canMergeSegments(Layout layout) {
+        if (layout.getSegments().size() == 1) {
+            return false;
+        } else {
+            int firstSegmentIndex = 0;
+            int secondSegmentIndex = 1;
+            return Sets.difference(
+                    layout.getSegments().get(secondSegmentIndex).getAllLogServers(),
+                    layout.getSegments().get(firstSegmentIndex).getAllLogServers()).isEmpty();
+        }
+    }
+
+    /**
+     * Returns true if any layout restoration action is needed. A restoration action is needed
+     * when any server can merge the segments or a current server can restore a layout redundancy.
+     *
+     * @param layout Current layout.
+     * @param server Current server.
+     * @return True, if restoration action is needed.
+     */
+    public static boolean canRestoreRedundancyOrMergeSegments(Layout layout, String server) {
+        return RedundancyCalculator.canRestoreRedundancy(layout, server) ||
+                RedundancyCalculator.canMergeSegments(layout);
+    }
+
+    /**
+     * Given a layout, and global trim mark creates an initial list
+     * of non-empty and bounded transfer segments.
+     *
+     * @param layout   A current layout.
+     * @param trimMark A current global trim mark.
+     * @return A list of transfer segments.
+     */
+    public ImmutableList<TransferSegment> createStateList(Layout layout, long trimMark) {
```

Comment:
Dealing with the inverse feels less convoluted than this logic.

Why not filer on !segmentContainsServer then use the left most segment and the right most segment endpoints.

I can see why this is being done, but its unnecessary to have logic that manages disjoint segments for a particular host, since that pattern of segmentation cannot happen. Since when a layout is constructed, we either remove a node from all segments or we added it to all segments, so how can you have a pattern where [seg1] [seg2] [seg3], but the node only exists on 1 and 3, but not 2?